src/v/storage/disk_log_impl.cc

```diff
@@ -81,16 +84,20 @@ ss::future<> disk_log_impl::remove() {
         permanent_delete.emplace_back(
           remove_segment_permanently(s, "disk_log_impl::remove()"));
     }
-    // wait for all futures
-    return ss::when_all_succeed(
-             permanent_delete.begin(), permanent_delete.end())
-      .then([this]() {
-          vlog(stlog.info, "Finished removing all segments:{}", config());
-      })
-      .then([this] {
-          return _kvstore.remove(
-            kvstore::key_space::storage,
-            internal::start_offset_key(config().ntp()));
+
+    return _readers_cache->stop().then(
```

Comment:
this strikes me as problematic (e.g. unnecessary delays controlled by random readers) to wait on readers. we didn't before, right? so it seems that if the cache is really a cache and not an _owning_ set of readers then it is in principle unnecessary. I added some more comments about this to the cache::stop method in another patch.

Reply:
we waited  for all those readers before, segment close needs a write lock.

Reply:
the point was that it adds unnecessary coupling and code. when the cache is stopped it can arrange for all entry guards to become noops (breaking dependency) and then all of this code where we have to wrap shutdown sequences in the stop continuation of the cache reader would seem to go away.

Reply:
we will require this `finally()` call for each reader that is not in use, additionally when reader is returned from the cache is wrapped in `cached_reader`. It makes the `finally()` call to be noop